Cauli/Florets/Inspector/Record List/InspectorTableViewController.swift

```diff
@@ -78,26 +83,75 @@ internal class InspectorTableViewController: UITableViewController {
         let distanceToBottom = scrollView.contentSize.height - scrollView.frame.height - scrollView.contentOffset.y
         guard !scrolledToEnd, !isLoading, distanceToBottom < 100 else { return }
         isLoading = true
-        let newRecords = cauli.storage.records(InspectorTableViewController.recordPageSize, after: records.last)
+        let newRecords = cauli.storage.records(InspectorTableViewController.recordPageSize, after: dataSource.items.last)
         if newRecords.isEmpty {
             isLoading = false
             scrolledToEnd = true
         } else if #available(iOS 11, *) {
             tableView.performBatchUpdates({
-                let indexPaths = (records.count..<(records.count + newRecords.count)).map { IndexPath(row: $0, section: 0) }
+                let numberOfExistingRecords = dataSource.filteredItems.count
+                let numberOfAddedRecords = dataSource.filter(records: newRecords).count
+                let indexPaths = (numberOfExistingRecords..<(numberOfExistingRecords + numberOfAddedRecords)).map { IndexPath(row: $0, section: 0) }
                 tableView.insertRows(at: indexPaths, with: .bottom)
-                records.append(contentsOf: newRecords)
+                dataSource.append(records: newRecords)
             }, completion: { [weak self] _ in
                 self?.isLoading = false
             })
         } else {
-            let indexPaths = (records.count..<(records.count + newRecords.count)).map { IndexPath(row: $0, section: 0) }
+            let numberOfExistingRecords = dataSource.filteredItems.count
+            let numberOfAddedRecords = dataSource.filter(records: newRecords).count
+            let indexPaths = (numberOfExistingRecords..<(numberOfExistingRecords + numberOfAddedRecords)).map { IndexPath(row: $0, section: 0) }
             tableView.beginUpdates()
             tableView.insertRows(at: indexPaths, with: .bottom)
-            records.append(contentsOf: newRecords)
+            dataSource.append(records: newRecords)
```

Comment:
Now that I am reading this, there is quite some code duplication going on here in these lines. The only difference is the `performBatchUpdates` vs the `beginUpdates` + `andUpdates`. Do you think we can find a nice way to deduplicate some lines here?

Reply:
We could do something like this, not sure if it's better:

```swift
        guard !newRecords.isEmpty else {
            isLoading = false
            scrolledToEnd = true
            return
        }
        let numberOfExistingRecords = dataSource.filteredItems.count
        let numberOfAddedRecords = dataSource.filter(records: newRecords).count
        let indexPaths = (numberOfExistingRecords..<(numberOfExistingRecords + numberOfAddedRecords)).map { IndexPath(row: $0, section: 0) }
        if #available(iOS 11, *) {
            tableView.performBatchUpdates({
                tableView.insertRows(at: indexPaths, with: .bottom)
                dataSource.append(records: newRecords)
            }, completion: { [weak self] _ in
                self?.isLoading = false
            })
        } else {
            tableView.beginUpdates()
            tableView.insertRows(at: indexPaths, with: .bottom)
            dataSource.append(records: newRecords)
            tableView.endUpdates()
            self.isLoading = false
        }
```

Reply:
Now that I am thinking of, is that maybe something that should be within the Datasource? So for example from the call site:

`datasource.append(records: newRecords, to: tableView)`

and then the implementation will perform more or less what we are currently doing here. This way we don't have these two public functions on the datasource `append` and `filter` but just the new append function.

About the question of reducing code duplication, what do you think about something like this:

```swift

// in the datasource

private func performBatchUpdate(in tableView: UITableView, closure: ()->(), completion: (Bool)->()) {
    if #available(iOS 11, *) {
        tableView.performBatchUpdates(closure, completion: completion)
    } else {
        tableView.beginUpdates()
        closure()
        tableView.endUpdates()
        completion()
    }
}
```

Reply:
Sounds like a clean approach. I will implement it then you can look over it again.

Reply:
Done. I think I like it, only the completion block chaining doesn't feel that good. Do you have a better idea than that?

Reply:
I like the changes. I am not sure what you mean by "completion block chaining". Do you mean the `}, completion: completion)` in line 88 in the `InspectorTableViewDatasource.swift`? If so, I am happy with that.

Reply:
I mean that the completion block is only defined in some methods to be passed to the next one. It felt like a cheap trick, "oh, I need the block here, let's pass it through" ..

Reply:
I don't think that is an anti pattern. I like it this way!